realm/realm-library/src/androidTestObjectServer/kotlin/io/realm/mongodb/RemoteMongoCollectionTest.kt

```diff
@@ -0,0 +1,248 @@
+        with(getCollectionInternal(COLLECTION_NAME)) {
+            val doc1 = Document("hello", "world1")
+            val doc2 = Document("hello", "world2")
+            val doc3 = Document("hello", "world3")
+
+            // Test findOne() on empty collection with no filter and no options
+            assertNull(findOne().blockingGetResult())
+
+            // Insert a document into the collection
+            insertOne(doc1).blockingGetResult()
+            assertEquals(1, count().blockingGetResult())
+
+            // Test findOne() with no filter and no options
+            assertEquals(doc1, findOne().blockingGetResult()!!.withoutId())
+
+            // Test findOne() with filter that does not match any documents and no options
+            assertNull(findOne(Document("hello", "worldDNE")).blockingGetResult())
+
+            // Insert 2 more documents into the collection
+            insertMany(listOf(doc2, doc3)).blockingGetResult()
+            assertEquals(3, count().blockingGetResult())
+
+            // FIXME: add these when all findOne methods are ready and the parser works
+            // test findOne() with projection and sort options
+//            val projection = Document("hello", "1")
+//            projection["_id"] = 0
+//            val options = RemoteFindOptions()
+//                    .limit(2)
+//                    .projection(projection)
+//                    .sort(Document("hello", "1"))
+//            assertEquals(findOne(Document(), options).blockingGetResult(), doc1.withoutId())
+
+            // FIXME: adapt to our framework
+//            val result3 = Tasks.await(coll.findOne(Document(), RemoteFindOptions()
+//                    .limit(2)
+//                    .projection(projection)
+//                    .sort(Document("hello", -1))))
+//            Assert.assertEquals(result3, withoutId(doc3))
+//
+//            // test findOne() properly fails
+//            try {
+//                Tasks.await(coll.findOne(Document("\$who", 1)))
+//                Assert.fail()
+//            } catch (ex: ExecutionException) {
+//                Assert.assertTrue(ex.cause is StitchServiceException)
+//                val svcEx = ex.cause as StitchServiceException
+//                assertEquals(StitchServiceErrorCode.MONGODB_ERROR, svcEx.errorCode)
+//            }
+        }
+    }
+
+    // FIXME: more to come
+
+    private fun getCollectionInternal(collectionName: String, javaClass: Class<Document>? = null): RemoteMongoCollection<Document> {
+        return when (javaClass) {
+            null -> database.getCollection(collectionName)
+            else -> database.getCollection(collectionName, javaClass)
+        }
+    }
+
+    private fun Document.withId(objectId: ObjectId? = null): Document {
+        return apply { this["_id"] = objectId ?: ObjectId() }
+    }
+
+    private fun Document.withoutId(): Document {
+        return apply { remove("_id") }
+    }
+
+    private companion object {
```

Comment:
True, and that's how I would normally do it for Java classes, but I am following Kotlin's style guide on companion objects: https://kotlinlang.org/docs/reference/coding-conventions.html#source-code-organization

Reply:
Fair enough. I wonder if we should have these in companion objects in that case or if it would make sense to convert them to `const val`s at the top of the file?

Reply:
The rule is companion objects should be used if the constants declared in them are class-scoped, and top-level constants should be used if they have a wider scope. I could see some of these might be used in the future elsewhere outside this test, but for now, they belong in this test. This could easily piggyback on using base classes for certain tests and such.